Adding a working example of docker linked mysql container for issue #887

ianblenke opened the pull request.

dsander commented:
I am using docker-machine and docker-compose 1.3.1 on OSX, installed via homebrew and docker-compose up (the command is still named fig) runs fine, but I can not connect to the huginn docker container.

Output of `docker ps`
```
CONTAINER ID        IMAGE               COMMAND                CREATED              STATUS              PORTS                    NAMES
52e88807a66d        docker_huginn       "/scripts/init"        About a minute ago   Up About a minute   0.0.0.0:5000->5000/tcp   docker_huginn_1
bbcd70327b75        mysql               "/entrypoint.sh mysq   About a minute ago   Up About a minute   3306/tcp                 docker_mysql_1
```

docker-machine ls
```
NAME   ACTIVE   DRIVER       STATE     URL                         SWARM
dev    *        virtualbox   Running   tcp://192.168.99.100:2376
```

I am trying to open huginn via `http://192.168.99.100:5000/`

dsander commented:
Never mind, figured that one out its still listening to port 3000 because it does not yet have your port fix.

dsander commented on docker/README.md:
```diff
@@ -51,13 +51,15 @@ Simple stand-alone usage:
         -e HUGINN_MYSQL_DATABASE=huginn \
         -e HUGINN_MYSQL_USER=huginn \
         -e HUGINN_MYSQL_PASSWORD=somethingsecret \
         -e HUGINN_MYSQL_ROOT_PASSWORD=somethingevenmoresecret \
```
I needed to change those environment variables to the ones you used in the `docker-compose.yml`, am I missing something?

cantino replied:
Should this be

```
docker run  --name mysql \
    -e MYSQL_ROOT_PASSWORD=myrootpassword \
    -e MYSQL_DATABASE=huginn \
    -e MYSQL_USER=huginn \
    -e MYSQL_PASSWORD=somethingsecret \
    mysql
```

?

ianblenke replied:
It should actually be HUGGIN_ prefixed due to the logic here:

https://github.com/cantino/huginn/blob/master/docker/scripts/init#L29

This:

```
grep = /app/.env.example | sed -e 's/^#\([^ ]\)/\1/' | grep -v -e '^#' | cut -d= -f1 | \
  while read var ; do
    eval "echo \"$var=\\\"\${HUGINN_$var:-\$$var}\\\"\""
  done | grep -v -e ^= > /app/.env
```

It looks for any HUGGINN_ prefixed variable and uses that to replace the environment sourced from the .env.example.

Think of it as an environment override.

dsander commented on docker/README.md:
```diff
@@ -51,13 +51,15 @@ Simple stand-alone usage:
         -e HUGINN_MYSQL_USER=huginn \
         -e HUGINN_MYSQL_PASSWORD=somethingsecret \
         -e HUGINN_MYSQL_ROOT_PASSWORD=somethingevenmoresecret \
         cantino/huginn
```
Shouldn't that be `mysql`?

dsander replied:
Still pretty sure it should be `mysql` or am I missing something?

ianblenke replied:
Yep, that should be mysql. You're not missing anything. Updated.

dsander commented on docker/README.md:
```diff
@@ -51,13 +51,15 @@ Simple stand-alone usage:
+        -p 5000:5000 \
         -e HUGINN_DATABASE_NAME=huginn \
         -e HUGINN_DATABASE_USER=huginn \
         -e HUGINN_DATABASE_PASSWORD=somethingsecret \
```
I had to use the mysql root password here, are both needed or just the root password?

ianblenke replied:
The `HUGINN_DATABASE_PASSWORD` is being used to override the `DATABASE_PASSWORD`
The `HUGINN_DATABASE_USERNAME` should be used to override the DATABASE_USERNAME`

I think it is defaulting the `DATABASE_USERNAME` to be `root`.

I also think you are specifying `HUGINN_DATABASE_USER` which would be overriding something called `DATABASE_USER`, which isn't used anywhere.

I've comitted a fix for the README.md to reflect this.

cantino commented:
Thanks @ianblenke!

@dsander, I'll let you be the reviewer on this PR.  I'm traveling, and don't know Docker well.

cantino commented:
Is this good to go @dsander and @ianblenke?

dsander commented:
@ianblenke Could you rebase this branch? I would like to check the `fig` and `README` examples again.

shaunaa126 commented:
@ianblenke @dsander @cantino I have been unsuccessful in trying to link Huginn to an external mySql container based on the original instructions. I can try again with these changes committed if you need someone to test. I do have a question.. based on the instructions in the readme, we are creating a mysql container based on the cantino/huginn image... can we instead create it from an official mysql image? Similar to the instructions here: https://registry.hub.docker.com/u/andrewcurioso/huginn/

dsander commented:
@shaunaa126 That should already work using the following commands. Not sure if that is reflected in the current readme though.

```
docker run  --name mysql \
    -e MYSQL_ROOT_PASSWORD=myrootpassword \
    -e MYSQL_DATABASE=huginn \
    -e MYSQL_USER=huginn \
    -e MYSQL_PASSWORD=somethingsecret \
    mysql
docker run  --name huginn \
    --link mysql \
    -p 3000:3000 \
    -e HUGINN_DATABASE_NAME=huginn \
    -e HUGINN_DATABASE_USERNAME=huginn \
    -e HUGINN_DATABASE_PASSWORD=somethingsecret \
    cantino/huginn
```

cantino commented:
@shaunaa126 Is the command @dsander listed different from the one you were trying?

shaunaa126 commented:
@cantino Yes, these commands are different from the ones you posted in Github/Docker Hub. The commands @dsander posted references the mysql image and correctly links it to the Huginn image. I have verified it works! Thank you both!

CONTAINER ID        IMAGE               COMMAND                CREATED              STATUS              PORTS                              NAMES
422d8533dc0f        shaunaa126/huginn   "/scripts/init"        41 seconds ago       Up 40 seconds       0.0.0.0:3000->3000/tcp, 5000/tcp   huginn              
1a25a46e819e        mysql               "/entrypoint.sh mysq   About a minute ago   Up About a minute   3306/tcp                           mysql

cantino commented:
@shaunaa126 Would you mind sending a PR that updates `docker/README.md` with the settings that worked for you?  If not, I'm happy to fix it.

cantino commented on docker/README.md:
```diff
@@ -51,21 +51,32 @@ Simple stand-alone usage:
-    docker run --rm --name huginn --link newcentury_mysql:MYSQL -p 5000:5000 \
+    docker run --rm --name huginn \
+        --link mysql \
+        -p 5000:5000 \
```
should this be 3000:3000?

cantino commented:
Wait, sorry, I think that's what this PR already is.

cantino commented:
@ianblenke @icblenke Would you have a chance to update this?  If not, @dsander or I can cut a new branch.

cantino commented:
@ianblenke Is this done?

ianblenke commented:
Apparently the mentions of port 5000 are confusing people. I am refactoring the branch for this PR to mention port 3000 as before to remediate this confusion.

ianblenke commented:
@dsander @cantino please review this PR one last time to make sure switching out mentions from port 5000 to port 3000 makes sense as the final fix.

I need a +1 so I can merge this and close out this issue.

ianblenke commented:
Issue #895 should be fixed by the latest update to this PR, replacing mention of port 5000 with port 3000.

When we merge this, we should be able to close out issue #895 as well.

cantino commented:
Thanks @ianblenke!  @dsander, would you mind reviewing?

dsander commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+mysqldata:
+  image: mysql
+  command: true
```
I was getting `json: cannot unmarshal bool into Go value of type string`

Changing it to `/bin/true` seems to have fixed it.

dsander commented on docker/README.md:
```diff
@@ -47,25 +47,36 @@ The CMD launches Huginn via the scripts/init script. This may become the ENTRYPO
 To link to another mysql container, for example:
 
-    docker run --rm --name newcentury_mysql -p 3306 \
+    docker run --rm --name mysql \
```
How about naming the container `mysql_huginn` or something like it to avoid conflicts with other mysql containers?

ianblenke commented:
with the merge of #983, this PR now conflicts rather badly.

ianblenke commented on docker/scripts/init:
```diff
@@ -147,19 +147,46 @@ cat <<FOREMAN > /tmp/foreman.sh
-[ -z "\${DO_NOT_MIGRATE}" ] && sudo -u huginn -EH bundle exec rake db:migrate
-[ -z "\${DO_NOT_SEED}" ] && sudo -u huginn -EH bundle exec rake db:seed
+if [ -z "\${DO_NOT_MIGRATE}" ]; then
+ sudo -u huginn -EH bundle exec rake db:migrate
```
@dsander, @cantino, should we really be doing a `rake db:migrate` here? Isn't it more appropriate to do a `rake db:schema:load`? It's not like we really want to update the `db/schema.rb`, we just want to load the `db/schema.rb` that is already in the project.

Or am I just picking nits?

dsander replied:
I think both would work, but I have never used `db:schema:load` apart from rspec which uses it to create the test database afaik.
Since the migration will once run once per container it probably does not make a huge difference.

ianblenke commented:
After thoroughly reviewing the comments made thus far, and addressing the concerns, the only remaining "issue" that I just identified is wether to change `rake db:migrate` to `rake db:schema:load`.

Honestly, I'm pretty sure this is as good as this PR really needs to get. Can I just get a +1 to merge it so we can close this silly thing?

dsander commented on docker/README.md:
```diff
@@ -51,21 +51,32 @@ Simple stand-alone usage:
         -e HUGINN_MYSQL_DATABASE=huginn \
         -e HUGINN_MYSQL_USER=huginn \
         -e HUGINN_MYSQL_PASSWORD=somethingsecret \
         -e HUGINN_MYSQL_ROOT_PASSWORD=somethingevenmoresecret \
```
Those lines need to read
```
        -e MYSQL_DATABASE=huginn \
        -e MYSQL_USER=huginn \
        -e MYSQL_PASSWORD=somethingsecret \
        -e MYSQL_ROOT_PASSWORD=somethingevenmoresecret \
````

dsander commented:
Thanks a lot @ianblenke! Apart for the environment variables for the linked mysql container everything works perfectly. :+1:

cantino commented:
Is stuff broken until we merge this?  If so, let's make that final change and merge.

cantino commented:
Thanks @ianblenke!  Hows this look @dsander?

dsander commented:
Perfect :+1:

cantino commented:
Awesome, go ahead and merge @ianblenke!  And thanks!
